### PR TITLE
Fix for GHSA matching and multiple fix versions

### LINKED ISSUE #633: npm packages are not matching from GHSA sources
**Version of Anchore Engine and Anchore CLI if applicable**: current tip of master, and I believe this has been a problem for the GHSA source since it was introduced

**What happened**:

```python
            elif self.pkg_type in ['npm', 'js']:
                likematch = '%js%'
                do_langscan = True
```

Will not match against the namespace in GHSA, which is `"github:npm"`. Other languages might be affected as well, so all need to be double-checked (e.g. Ruby)

Tangentially, the `Fixed In` reporting is not entirely accurate as many affected versions can have many "fixed in" values:

```
Vulnerability ID           Package         Severity        Fix          CVE Refs             Vulnerability URL                                        Type        Feed Group        Package Path                         
GHSA-pp7h-53gx-mx7r        bl-2.2.0        High            1.2.3        CVE-2020-8244        https://github.com/advisories/GHSA-pp7h-53gx-mx7r        npm         github:npm        /node_modules/bl/package.json   
```

The above does not represent what is displayed here https://github.com/advisories/GHSA-pp7h-53gx-mx7r

### LINKED ISSUE #639: Vulnerability fix version isn't correct for vulnerabilities that have multiple fixes available
Some records, like GitHub Security Advisories will include multiple fix versions for different ranges of a library. The system currently (as of v0.8.1) picks the first match on the name, rather than the with the range that matches the actual found artifact in the image.

### PR DESCRIPTION
**Pull request by zhill**
Adds fixes for npm and gem matching with GHSA records and gets correct fix version when multiple fixes available.

Fixes #633 
Fixes #639

### CODE REVIEW

**Comment by dakaneye** [comment]
LGTM!

File: tests/unit/anchore_engine/services/policy_engine/engine/test_vulnerability_matches.py [hunk 1]
```diff
@@ -63,6 +78,61 @@ def vulnerability_with_fix(empty_vulnerability):
...
+    f.parent = fixed_vuln
+    f.include_later_versions = False
+    f.epochless_version = f.version
+    f.fix_metadata = {'first_patched_version': '1.1.2'}
+    f.created_at = datetime.datetime.now()
+    f.updated_at = datetime.datetime.now()
+    f.fix_observed_at = f.updated_at
+    fixed_vuln.fixed_in = [f]
+
+    f = FixedArtifact()
```
Comment (alfredodeza) [inline tests/unit/anchore_engine/services/policy_engine/engine/test_vulnerability_matches.py:119]: this `f` instance doesn't look like it is used in `fixed_vuln`, that is probably intentional, but I am not following how that is working to get two fixed artifacts if the `empty_semver_vulnerability` is only used with `fixed_vuln`

Comment (zhill) [reply tests/unit/anchore_engine/services/policy_engine/engine/test_vulnerability_matches.py:119]: The parent relationship connects it up in the fixed_in collection (`f.parent = fixed_vuln`) due to how SQLAlchemy classes work with the relationship definition. Line 117 is actually redundant and can be removed.

File: anchore_engine/db/entities/policy_engine.py [hunk 2]
```diff
@@ -1883,30 +1883,51 @@ class ImagePackageVulnerability(Base):
...
+        candidates = self.fix_candidates()
+        candidate_count = len(candidates) if candidates else 0
+        if candidate_count == 0:
+            return None
+
+        fixed_in = None
+
+        # candidate_count >= 1
+        # Do version checks. This package must be affected by the range but not fixed.
+        matched = [x for x in candidates if x.match_but_not_fixed(self.package)]
```
Comment (nightfurys) [inline anchore_engine/db/entities/policy_engine.py:1916]: Yep, prior this change `match_but_not_fixed` is used for finding matches but was not used for looking up the associated fixed in version (or no fix). This is exactly the change I was thinking of as well. lgtm

File: tests/unit/anchore_engine/services/policy_engine/engine/test_vulnerability_matches.py (hunk 1 above)
Comment (alfredodeza) [reply tests/unit/anchore_engine/services/policy_engine/engine/test_vulnerability_matches.py:119]: Ahhh that is what I was hoping for, but wasn't sure that we were dealing with a model here.

**Pull request merged**